## Capstone Project – The Battle of Neighborhoods | Finding a Better Place in Scarborough, Toronto

### 1.Installing and Importing Python Libraries and Dependencies

In [2]:
!pip install geocoder
!pip install folium

     |████████████████████████████████| 98 kB 8.3 MB/s  eta 0:00:01
     |████████████████████████████████| 94 kB 5.5 MB/s  eta 0:00:01


##### import libraries

In [3]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print("All Required Libraries Imported!")

All Required Libraries Imported!


### 2. Data Extraction and Cleaning

##### Using BeautifulSoup Scraping List of Postal Codes of Given Wikipedia Page. 
 #### Link: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [68]:

url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
extracting_data = requests.get(url).text
wiki_data = BeautifulSoup(extracting_data, 'lxml')

column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

content = wiki_data.find('div', class_='mw-parser-output')
table = content.table.tbody
postcode = 0
borough = 0
neighborhood = 0

for tr in table.find_all('tr'):
    i = 0
    for td in tr.find_all('td'):
        if i == 0:
            postcode = td.text
            i = i + 1
        elif i == 1:
            borough = td.text
            i = i + 1
        elif i == 2: 
            neighborhood = td.text.strip('\n').replace(']','')
    toronto = toronto.append({'Postalcode': postcode,'Borough': borough,'Neighborhood': neighborhood},ignore_index=True)

In [69]:
# clean dataframe 
toronto = toronto[toronto.Borough!='Not assigned']
toronto = toronto[toronto.Borough!= 0]
toronto.reset_index(drop = True, inplace = True)
i = 0
for i in range(0,toronto.shape[0]):
    if toronto.iloc[i][2] == 'Not assigned':
        toronto.iloc[i][2] = toronto.iloc[i][1]
        i = i+1
df = toronto.groupby(['Postalcode','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df.head()

df = df.dropna()
empty = 'Not assigned'
df = df[(df.Postalcode != empty ) & (df.Borough != empty) & (df.Neighborhood != empty)]


def neighborhood_list(grouped):    
    return ', '.join(sorted(grouped['Neighborhood'].tolist()))
                    
grp = df.groupby(['Postalcode', 'Borough'])
df_2 = grp.apply(neighborhood_list).reset_index(name='Neighborhood')

In [70]:
def get_latilong(postal_code):
    lati_long_coords = None
    while(lati_long_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(postal_code))
        lati_long_coords = g.latlng
    return lati_long_coords
    
get_latilong('M4G')

[43.709020000000066, -79.36348999999996]

In [71]:

# Retrieving Postal Code Co-ordinates
postal_codes = df_2['Postalcode']    
coords = [ get_latilong(postal_code) for postal_code in postal_codes.tolist() ]

In [72]:
# Adding Columns Latitude & Longitude
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])
df_2['Latitude'] = df_coords['Latitude']
df_2['Longitude'] = df_coords['Longitude']

In [73]:

df_2[df_2.Postalcode == 'M5G']

Empty DataFrame
Columns: [Postalcode, Borough, Neighborhood, Latitude, Longitude]
Index: []

In [74]:
df_2.head(10)

Postalcode         Borough                                     Neighborhood  \
0      M1A\n  Not assigned\n                                   Not assigned\n   
1      M1B\n   Scarborough\n                                   Malvern, Rouge   
2      M1C\n   Scarborough\n           Rouge Hill, Port Union, Highland Creek   
3      M1E\n   Scarborough\n                Guildwood, Morningside, West Hill   
4      M1G\n   Scarborough\n                                           Woburn   
5      M1H\n   Scarborough\n                                        Cedarbrae   
6      M1J\n   Scarborough\n                              Scarborough Village   
7      M1K\n   Scarborough\n      Kennedy Park, Ionview, East Birchmount Park   
8      M1L\n   Scarborough\n                  Golden Mile, Clairlea, Oakridge   
9      M1M\n   Scarborough\n  Cliffside, Cliffcrest, Scarborough Village West   

   Latitude  Longitude  
0  43.64869  -79.38544  
1  43.81139  -79.19662  
2  43.78574  -79.15875  
3  43.76575  -79.17470  
4  43.76812  -79.21761  
5  43.76944  -79.23892  
6  43.74446  -79.23117  
7  43.72582  -79.26461  
8  43.71289  -79.28506  
9  43.72360  -79.23496

In [75]:
address = 'Scarborough,Toronto'

geolocator = Nominatim(user_agent='ruozhou')
location = geolocator.geocode(address)
latitude_x = location.latitude
longitude_y = location.longitude
print('The Geograpical Co-ordinate of Seattle,Washington are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Seattle,Washington are 43.7729744, -79.2576479.


### 3. Map of Scarborough

In [76]:
map_Scarborough = folium.Map(location=[latitude_x, longitude_y], zoom_start=10)

for lat, lng, nei in zip(df_2['Latitude'], df_2['Longitude'], df_2['Neighborhood']):
    
    label = '{}'.format(nei)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Scarborough)  
    
map_Scarborough

In [77]:

address = 'Scarborough,Toronto'

geolocator = Nominatim(user_agent='ruozhou')
location = geolocator.geocode(address)
latitude_n1 = location.latitude
longitude_n1 = location.longitude
print('The Geograpical Co-ordinate of Neighborhood_1 are {}, {}.'.format(latitude_x, longitude_y))

The Geograpical Co-ordinate of Neighborhood_1 are 43.7729744, -79.2576479.


In [78]:

# @hiddel_cell
CLIENT_ID = 'BNIMWF2TQGJYNATHDBI0QHRCRAXDI3ZGZZQOYSVHN41C1251' # my Foursquare ID
CLIENT_SECRET = 'FQUIVY0VI5IEJKIWS2HDUDEDXBUG45IJWOW2LKSM3JLAQQTZ' # my Foursquare Secret
VERSION = '20180605'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: '+CLIENT_ID)
print('CLIENT_SECRET: '+CLIENT_SECRET)

Your credentails:
CLIENT_ID: BNIMWF2TQGJYNATHDBI0QHRCRAXDI3ZGZZQOYSVHN41C1251
CLIENT_SECRET: FQUIVY0VI5IEJKIWS2HDUDEDXBUG45IJWOW2LKSM3JLAQQTZ


In [79]:
radius = 700 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude_n1, 
   longitude_n1, 
    radius, 
   LIMIT)
results = requests.get(url).json()

In [80]:
venues=results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues)
nearby_venues.columns

Index(['referralId', 'reasons.count', 'reasons.items', 'venue.id',
       'venue.name', 'venue.location.address', 'venue.location.crossStreet',
       'venue.location.lat', 'venue.location.lng',
       'venue.location.labeledLatLngs', 'venue.location.distance',
       'venue.location.postalCode', 'venue.location.cc', 'venue.location.city',
       'venue.location.state', 'venue.location.country',
       'venue.location.formattedAddress', 'venue.categories',
       'venue.photos.count', 'venue.photos.groups',
       'venue.location.neighborhood', 'venue.venuePage.id'],
      dtype='object')

In [81]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### 4. Nearby Venues/Locations

In [82]:
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues.head()

venue.name  \
0                  Disney Store   
1                       SEPHORA   
2  Coliseum Scarborough Cinemas   
3                Tommy Hilfiger   
4          American Eagle Store   

                                    venue.categories  venue.location.lat  \
0  [{'id': '4bf58dd8d48988d1f3941735', 'name': 'T...           43.775537   
1  [{'id': '4bf58dd8d48988d10c951735', 'name': 'C...           43.775017   
2  [{'id': '4bf58dd8d48988d17f941735', 'name': 'M...           43.775995   
3  [{'id': '4bf58dd8d48988d103951735', 'name': 'C...           43.776015   
4  [{'id': '4bf58dd8d48988d103951735', 'name': 'C...           43.776012   

   venue.location.lng  
0          -79.256833  
1          -79.258109  
2          -79.255649  
3          -79.257369  
4          -79.258334

### 5. Categories of Nearby Venues/Locations

In [83]:
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(5)

name        categories        lat        lng
0                  Disney Store  Toy / Game Store  43.775537 -79.256833
1                       SEPHORA    Cosmetics Shop  43.775017 -79.258109
2  Coliseum Scarborough Cinemas     Movie Theater  43.775995 -79.255649
3                Tommy Hilfiger    Clothing Store  43.776015 -79.257369
4          American Eagle Store    Clothing Store  43.776012 -79.258334

In [84]:
# Top 10 Categories
a=pd.Series(nearby_venues.categories)
a.value_counts()[:10]

Clothing Store            8
Restaurant                5
Coffee Shop               5
Sandwich Place            2
Pharmacy                  2
Department Store          2
Gas Station               2
Bar                       1
Sporting Goods Shop       1
Furniture / Home Store    1
Name: categories, dtype: int64

In [85]:
def getNearbyVenues(names, latitudes, longitudes, radius=700):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # making GET request
        venue_results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in venue_results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [86]:
# Nearby Venues
Scarborough_venues = getNearbyVenues(names=df_2['Neighborhood'],
                                   latitudes=df_2['Latitude'],
                                   longitudes=df_2['Longitude']
                                  )

Not assigned

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Not assigned

Parkwoods
Don Mills
Don Mills
Not assigned

Not assigned

Bathurst Manor, Wilson Heights, Downsview North
No

In [87]:
print('There are {} Uniques Categories.'.format(len(Scarborough_venues['Venue Category'].unique())))
Scarborough_venues.groupby('Neighborhood').count().head()

There are 307 Uniques Categories.


Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                           21   
Alderwood, Long Branch                                               6   
Bathurst Manor, Wilson Heights, Downsview North                     15   
Bayview Village                                                      6   
Bedford Park, Lawrence Manor East                                   26   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                            21   
Alderwood, Long Branch                                                6   
Bathurst Manor, Wilson Heights, Downsview North                      15   
Bayview Village                                                       6   
Bedford Park, Lawrence Manor East                                    26   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                           21              21   
Alderwood, Long Branch                               6               6   
Bathurst Manor, Wilson Heights, Downsview North     15              15   
Bayview Village                                      6               6   
Bedford Park, Lawrence Manor East                   26              26   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                     21   
Alderwood, Long Branch                                         6   
Bathurst Manor, Wilson Heights, Downsview North               15   
Bayview Village                                                6   
Bedford Park, Lawrence Manor East                             26   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                    21  
Alderwood, Long Branch                                        6  
Bathurst Manor, Wilson Heights, Downsview North              15  
Bayview Village                                               6  
Bedford Park, Lawrence Manor East                            26

#### One Hot Encoding of Features

In [88]:
# one hot encoding
Scarborough_onehot = pd.get_dummies(Scarborough_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Scarborough_onehot['Neighborhood'] = Scarborough_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Scarborough_onehot.columns[-1]] + list(Scarborough_onehot.columns[:-1])
Scarborough_onehot = Scarborough_onehot[fixed_columns]
Scarborough_grouped = Scarborough_onehot.groupby('Neighborhood').mean().reset_index()
Scarborough_onehot.head(5)

Zoo Exhibit  Accessories Store  African Restaurant  Airport  \
0            0                  0                   0        0   
1            0                  0                   0        0   
2            0                  0                   0        0   
3            0                  0                   0        0   
4            0                  0                   0        0   

   American Restaurant  Animal Shelter  Antique Shop  Aquarium  Art Gallery  \
0                    0               0             0         0            0   
1                    0               0             0         0            0   
2                    0               0             0         0            0   
3                    0               0             0         0            0   
4                    0               0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Dealership  Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  \
0                0              0                0          0           0   
1                0              0                0          0           0   
2                0              0                0          0           0   
3                0              0                0          0           0   
4                0              0                0          0           0   

   Badminton Court  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0                0           0       0     0    0               0   
1                0           0       0     0    0               0   
2                0           0       0     0    0               0   
3                0           0       0     0    0               0   
4                0           0       0     0    0               0   

   Baseball Stadium  Basketball Stadium  Beach  Beach Bar  Beer Bar  \
0                 0                   0      0          0         0   
1                 0                   0      0          0         0   
2                 0                   0      0          0         0   
3                 0                   0      0          0         0   
4                 0                   0      0          0         0   

   Beer Store  Belgian Restaurant  Big Box Store  Bike Shop  Bistro  \
0           0                   0              0          0       0   
1           0                   0              0          0       0   
2           0                   0              0          0       0   
3           0                   0              0          0       0   
4           0                   0              0          0       0   

   Board Shop  Boat or Ferry  Bookstore  Botanical Garden  Boutique  \
0           0              0          0                 0         0   
1           0              0          0                 0         0   
2           0              0          0                 0         0   
3           0              0          0                 0         0   
4           0              0          0                 0         0   

   Breakfast Spot  Brewery  Bridge  Bubble Tea Shop  Buffet  Building  \
0               0        0       0                0       0         0   
1               0        0       0                0       0         0   
2               0        0       0                0       0         0   
3               0        0       0                0       0         0   
4               0        0       0                0       0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0        

In [89]:

num_top_venues = 5
for hood in Scarborough_grouped['Neighborhood']:
    print("---- "+hood+" ----")
    temp =Scarborough_grouped[Scarborough_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---- Agincourt ----
                venue  freq
0       Shopping Mall  0.10
1  Chinese Restaurant  0.10
2            Pharmacy  0.05
3        Skating Rink  0.05
4                Bank  0.05


---- Alderwood, Long Branch ----
         venue  freq
0          Gym  0.17
1  Gas Station  0.17
2          Pub  0.17
3  Pizza Place  0.17
4  Coffee Shop  0.17


---- Bathurst Manor, Wilson Heights, Downsview North ----
                       venue  freq
0                       Park  0.13
1                Pizza Place  0.13
2                Coffee Shop  0.13
3  Middle Eastern Restaurant  0.07
4               Intersection  0.07


---- Bayview Village ----
              venue  freq
0  Asian Restaurant  0.17
1              Park  0.17
2       Flower Shop  0.17
3             Trail  0.17
4           Dog Run  0.17


---- Bedford Park, Lawrence Manor East ----
                venue  freq
0  Italian Restaurant  0.08
1      Sandwich Place  0.08
2          Hobby Shop  0.08
3         Coffee Shop  0.08
4   Indian 

In [90]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Most Common venues near neighborhood

In [91]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Scarborough_grouped['Neighborhood']

for ind in np.arange(Scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch    Athletics & Sports   
2  Bathurst Manor, Wilson Heights, Downsview North           Pizza Place   
3                                  Bayview Village               Dog Run   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall       Badminton Court        Breakfast Spot   
1                   Pub           Gas Station                   Gym   
2                  Park           Coffee Shop           Men's Store   
3           Flower Shop                  Park                 Trail   
4            Hobby Shop        Sandwich Place    Italian Restaurant   

  5th Most Common Venue     6th Most Common Venue      7th Most Common Venue  \
0          Skating Rink                      Café  Latin American Restaurant   
1           Pizza Place               Coffee Shop                Yoga Studio   
2         Deli / Bodega  Mediterranean Restaurant           Sushi Restaurant   
3           Gas Station          Asian Restaurant                 Donut Shop   
4           Pizza Place                   Butcher                       Café   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0           Supermarket      Sushi Restaurant               Sandwich Place  
1           Dry Cleaner   Dumpling Restaurant  Eastern European Restaurant  
2   Fried Chicken Joint        Sandwich Place                 Intersection  
3           Dry Cleaner   Dumpling Restaurant  Eastern European Restaurant  
4          Liquor Store           Sports Club             Sushi Restaurant

#### K-Means Clustering Approach

In [92]:
# Using K-Means to cluster neighborhood into 3 clusters
Scarborough_grouped_clustering = Scarborough_grouped.drop('Neighborhood', 1)
kmeans = KMeans(n_clusters=3, random_state=0).fit(Scarborough_grouped_clustering)
kmeans.labels_

array([1, 1, 1, 2, 1, 1, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 2, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1,
       1, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       2, 1, 1, 1, 1, 1, 2, 2, 1, 1, 2], dtype=int32)

In [93]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Scarborough_merged =df_2.iloc[:16,:]

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Scarborough_merged = Scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Scarborough_merged.head()# check the last columns!

Postalcode         Borough                            Neighborhood  \
0      M1A\n  Not assigned\n                          Not assigned\n   
1      M1B\n   Scarborough\n                          Malvern, Rouge   
2      M1C\n   Scarborough\n  Rouge Hill, Port Union, Highland Creek   
3      M1E\n   Scarborough\n       Guildwood, Morningside, West Hill   
4      M1G\n   Scarborough\n                                  Woburn   

   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.64869  -79.38544               1           Coffee Shop   
1  43.81139  -79.19662               0           Zoo Exhibit   
2  43.78574  -79.15875               1                   Bar   
3  43.76575  -79.17470               2                  Park   
4  43.76812  -79.21761               2                  Park   

        2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                       Hotel                  Café            Restaurant   
1  Construction & Landscaping  Fast Food Restaurant           Flower Shop   
2           Fish & Chips Shop           Yoga Studio                  Farm   
3        Gym / Fitness Center        Gymnastics Gym    Athletics & Sports   
4          Chinese Restaurant  Fast Food Restaurant           Coffee Shop   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar                          Gym        Cosmetics Shop   
1           Flea Market                  Fish Market     Fish & Chips Shop   
2   Dumpling Restaurant  Eastern European Restaurant     Electronics Store   
3           Yoga Studio                      Dog Run            Donut Shop   
4           Yoga Studio                  Event Space           Dry Cleaner   

  8th Most Common Venue        9th Most Common Venue  \
0             Bookstore                 Concert Hall   
1   Filipino Restaurant                        Field   
2  Ethiopian Restaurant                  Event Space   
3           Dry Cleaner          Dumpling Restaurant   
4   Dumpling Restaurant  Eastern European Restaurant   

        10th Most Common Venue  
0                      Theater  
1                   Donut Shop  
2           Falafel Restaurant  
3  Eastern European Restaurant  
4            Electronics Store

#### Map of Clusters

In [100]:
kclusters = 10

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Postalcode    180 non-null    object 
 1   Borough       180 non-null    object 
 2   Neighborhood  180 non-null    object 
 3   Latitude      180 non-null    float64
 4   Longitude     180 non-null    float64
dtypes: float64(2), object(3)
memory usage: 7.2+ KB


In [95]:
# create map
map_clusters = folium.Map(location=[latitude_x, longitude_y], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
colors_array = cm.rainbow(np.linspace(0, 1, kclusters))
rainbow = [colors.rgb2hex(i) for i in colors_array]
print(rainbow)
# add markers to the map

markers_colors = []
for lat, lon, nei , cluster in zip(Scarborough_merged['Latitude'], 
                                   Scarborough_merged['Longitude'], 
                                   Scarborough_merged['Neighborhood'], 
                                   Scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(nei) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

['#8000ff', '#4856fb', '#10a2f0', '#2adddd', '#62fbc4', '#9cfba4', '#d4dd80', '#ffa256', '#ff562c', '#ff0000']


In [96]:
df1=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 0,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df2=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 1,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]
df3=Scarborough_merged.loc[Scarborough_merged['Cluster Labels'] == 2,Scarborough_merged.columns[[2] + list(range(5, Scarborough_merged.shape[1]))]]

In [99]:
Scarborough_Avg_HousingPrice=pd.DataFrame({"Neighborhood":df_2["Neighborhood"],"Average_Housing_Price":[335000.0,286600.0,175000.0,225900.0,219400.0,573900.0,225000.0,370500.0,370500.0,433500.0,279200.0,279200.0,225000.0,370500.,255400.0,433500.0,433500.0,435000.0,289500.0,265000.0,285900.0,239400.0,589900.0,295000.0,380500.0,378500.0,438500.0,229200.0,229200.0,365000.0,388500.,285400.0,493500.0,477500.0,378000.0,316600.0,195000.0,225900.0,219400.0,573900.0,367000.0,370500.0,370500.0,363500.0,279200.0,279200.0,271000.0,370500.,255400.0,383500.0,433500.0,335000.0,286600.0,185000.0,225900.0,219400.0,573900.0,329000.0,370500.0,370500.0,533500.0,279200.0,279200.0,375000.0,370500.,255400.0,493500.0,433500.0,335000.0,286600.0,165000.0,225900.0,219400.0,573900.0,425000.0,370500.0,370500.0,433500.0,279200.0,279200.0,195000.0,370500.,255400.0,403500.0,433500.0,335000.0,286600.0,187000.0,225900.0,219400.0,573900.0,325000.0,370500.0,370500.0,333500.0,279200.0,279200.0,289000.0,370500.,255400.0,413500.0,433500.0,254800.0]})

ValueError: array length 103 does not match index length 180

In [ ]:
Scarborough_Avg_HousingPrice.set_index('Neighborhood',inplace=True,drop=True)

In [ ]:
Scarborough_Avg_HousingPrice.plot(kind='bar',figsize=(24,18),alpha=0.75)

#### School Rating by Clusters

In [ ]:
clusters=pd.DataFrame({"Cluster1":df1["Neighborhood"],
                      "Cluster2":df2["Neighborhood"],
                      "Cluster4":df3["Neighborhood"]})
clusters = clusters.replace(np.nan, '', regex=True)

In [ ]:
new_Scarborough=Scarborough_merged.set_index("Neighborhood",drop=True)
#Source:https://www.greatschools.org
Scarborough_school_ratings=pd.DataFrame({"Neighborhood":df["Neighborhood"],
                                      "Top School Rating":[7,9,5,8,10,10,7,10,1,2,1,2,7,2,3,2,6,
                                                           5,4,8,9,9,6,6,4,5,4,6,8,10,8,9,6,2,
                                                           10,2,5,8,9,6,6,10,8,9,1,2,3,4,5,6,9,
                                                           8,5,9,6,9,6,4,8,10,2,5,6,3,9,8,7,
                                                           7,8,5,8,9,1,5,4,7,2,3,6,6,9,4,8,7,
                                                           4,8,9,2,6,4,7,5,10,4,6,8,9,7,5,6,5,8,7
                                                           ]})

In [ ]:
Scarborough_school_ratings.set_index('Neighborhood',inplace=True,drop=True)

In [ ]:
Scarborough_school_ratings.plot(kind='bar',figsize=(16,10),color='green',alpha=0.75);

#### Conclusion: In this project, using k-means cluster algorithm I separated the neighborhood into 10(Ten) different clusters and for 103 different lattitude and logitude from dataset, which have very-similar neighborhoods around them. Using the charts above results presented to a particular neighborhood based on average house prices and school rating have been made.